google/resource_compute_shared_vpc.go

```diff
@@ -0,0 +1,194 @@
+
+	"google.golang.org/api/compute/v1"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceComputeSharedVpc() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceComputeSharedVpcCreate,
+		Read:   resourceComputeSharedVpcRead,
+		Update: resourceComputeSharedVpcUpdate,
+		Delete: resourceComputeSharedVpcDelete,
+
+		Schema: map[string]*schema.Schema{
+			"host_project": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"service_projects": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceComputeSharedVpcCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	hostProject := d.Get("host_project").(string)
+	op, err := config.clientCompute.Projects.EnableXpnHost(hostProject).Do()
+	if err != nil {
+		return fmt.Errorf("Error enabling Shared VPC Host %q: %s", hostProject, err)
+	}
+
+	d.SetId(hostProject)
+
+	err = computeOperationWait(config, op, hostProject, "Enabling Shared VPC Host")
+	if err != nil {
+		d.SetId("")
+		return err
+	}
+
+	if v, ok := d.GetOk("service_projects"); ok {
+		serviceProjects := convertStringArr(v.(*schema.Set).List())
+		for _, project := range serviceProjects {
+			if err = enableResource(config, hostProject, project); err != nil {
+				return fmt.Errorf("Error enabling Shared VPC service project %q: %s", project, err)
+			}
+		}
+	}
+
+	return resourceComputeSharedVpcRead(d, meta)
+}
+
+func resourceComputeSharedVpcRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	hostProject := d.Get("host_project").(string)
+
+	project, err := config.clientCompute.Projects.Get(hostProject).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Project data for project %q", hostProject))
+	}
+
+	if project.XpnProjectStatus != "HOST" {
+		log.Printf("[WARN] Removing Shared VPC host resource %q because it's not enabled server-side", hostProject)
+		d.SetId("")
+	}
+
+	serviceProjects := []string{}
+	req := config.clientCompute.Projects.GetXpnResources(hostProject)
+	if err := req.Pages(context.Background(), func(page *compute.ProjectsGetXpnResources) error {
+		for _, xpnResourceId := range page.Resources {
+			if xpnResourceId.Type == "PROJECT" {
+				serviceProjects = append(serviceProjects, xpnResourceId.Id)
+			}
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("Error reading Shared VPC service projects for host %q: %s", hostProject, err)
+	}
+
+	d.Set("service_projects", serviceProjects)
+
+	return nil
+}
+
+func resourceComputeSharedVpcUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	hostProject := d.Get("host_project").(string)
+
+	if d.HasChange("service_projects") {
+		old, new := d.GetChange("service_projects")
+		// Helper to convert slice to map
+		m := func(vals []interface{}) map[string]struct{} {
```

Comment:
Consider moving this to our `utils.go` file.

Reply:
Done.